enginetest/join_planning_tests.go

```diff
@@ -1157,6 +1157,110 @@ join uv d on d.u = c.x`,
 					{6, 4, 6},
 				},
 			},
+			{
+				q:     "select * from vals join ranges on val >= min and val <= max where min >= 2",
+				types: []plan.JoinType{plan.JoinTypeInner},
+				exp: []sql.Row{
+					{2, 2, 4},
+					{3, 2, 4},
+					{3, 3, 5},
+					{4, 2, 4},
+					{4, 3, 5},
+					{4, 4, 6},
+					{5, 3, 5},
+					{5, 4, 6},
+					{6, 4, 6},
+				},
+			},
+			{
+				q:     "select * from vals join ranges on val between min and max where min >= 2 and max <= 5",
+				types: []plan.JoinType{plan.JoinTypeInner},
+				exp: []sql.Row{
+					{2, 2, 4},
+					{3, 2, 4},
+					{3, 3, 5},
+					{4, 2, 4},
+					{4, 3, 5},
+					{5, 3, 5},
+				},
+			},
+			{
+				q:     "select * from vals join (select max, min from ranges) ranges on val between min and max where min >= 2 and max <= 5",
+				types: []plan.JoinType{plan.JoinTypeRangeHeap},
+				exp: []sql.Row{
+					{2, 4, 2},
+					{3, 4, 2},
+					{3, 5, 3},
+					{4, 4, 2},
+					{4, 5, 3},
+					{5, 5, 3},
+				},
+			},
+			{
+				q:     "select * from vals join (select * from ranges where min >= 2 and max <= 5) ranges on val between min and max",
+				types: []plan.JoinType{plan.JoinTypeRangeHeap},
+				exp: []sql.Row{
+					{2, 2, 4},
+					{3, 2, 4},
+					{3, 3, 5},
+					{4, 2, 4},
+					{4, 3, 5},
+					{5, 3, 5},
+				},
+			},
+			{
+				q:     "select * from vals join (select * from ranges where min >= 2 and max <= 5 limit 1) ranges on val between min and max",
```

Comment:
i don't think this one is valid as a rangeHeap, should block this transform

Reply:
distinct is another one that might error, because we don't materialize it into the tree until after join planning